### PR TITLE
kernel: Timer behavioral testing

### PR DESCRIPTION
**Pull request by teburd**
Test and validate the behavior of a timer driver.

Takes a number of absolute timer cycle samples of a periodic timer then
calculates statistical mean, variance, stddev along with total drift over
the entire test time.

Fails on at least a few platforms for likely a variety of reasons. The default test bounds are arbitrarily chosen at the moment but seems reasonable (drift of less than 1000us over 10s, stddev of less than 10us).

Tested platforms, successes and failures

* intel adsp cavs25 passes (with fixes)
* nrf52840dk_nrf52840 fails (elapsed time is 60ms longer than expected, stddev fine)
* mimxrt1060_evk fails (elapsed time is 2s longer than expected, stddev fine)
* frdm_k64f passes
* nucleo_f446re passes
* tdk_robokit1 passes

Passers are all well under 1us stddev and +/- 100us from total elapsed time vs expected.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

### CODE REVIEW

**Comment by nashif** [comment]
> ERROR   - /home/nashif/zephyrproject/zephyr/tests/kernel/timer/timer_behavior/src/main.c: Found a test that does not start with test_

File: tests/kernel/timer/timer_behavior/src/main.c [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+#endif
+		periodic_data[periodic_idx + 1] = periodic_end;
+		periodic_idx++;
+	} else {
+		k_timer_stop(t);
+		k_sem_give(&periodic_sem);
+	}
+}
+
+ZTEST(timer_behavior, periodic_behavior)
```
Comment (nashif) [inline tests/kernel/timer/timer_behavior/src/main.c:37]: should be test_periodic_behavior

File: tests/kernel/timer/timer_behavior/Kconfig [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+source "Kconfig.zephyr"
+
+config TIMER_TEST_SAMPLES
+	int "The number of timer samples to gather for statistics"
+	default 10000
+   
+config TIMER_TEST_PERIOD
+	int "The number of microseconds to for the timer period"
+	default 1000
```
Comment (andyross) [inline tests/kernel/timer/timer_behavior/Kconfig:14]: Nothing wrong with specifying this in ms here, but when used at runtime you want to be converting this to an integer number of ticks before use and doing all the validation math relative to that number, not real time.  There are multiple platforms now (Nordic being the giant in the list) with hardware clocks that can't produce a 1ms delay exactly.

File: tests/kernel/timer/timer_behavior/src/main.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	for (int i = 0; i < CONFIG_TIMER_TEST_SAMPLES; i++) {
+		total_cycles += (periodic_data[i + 1] - periodic_data[i]);
+	}
+
+	double mean = (double)total_cycles / (double)CONFIG_TIMER_TEST_SAMPLES;
+
+	double variance = 0;
+
+	for (int i = 0; i < CONFIG_TIMER_TEST_SAMPLES; i++) {
+		double diff = (double)(periodic_data[i + 1] - periodic_data[i]) - mean;
```
Comment (andyross) [inline tests/kernel/timer/timer_behavior/src/main.c:67]: This doesn't handle wraparound of non-64-bit cycle counts correctly as it's doing the math in 64 bit precision (i.e. you want `0x00001000 - 0xfffff000` to produce 0x2000, but it actually gives you a huge 64 bit value instead).  My suggestion would be do do the interval subtraction in the k_timer callback instead, which can be guaranteed not to overflow as long as the interval isn't pathologically long.

Also, note that ARM SysTick is actually a **24 bit** hardware counter, so k_cycle_get_32() wraps at like 6 Hz on  the faster boards (There was an Atmel board that was susceptible to this, forget the name).

**Review by andyross — COMMENTED** [review]
Some notes.  I love the idea here.  I'm curious about the intel_adsp issues you're seeing.  The timer hardware has been banged pretty hard there (e.g. there's a calibration loop between CCOUNT and HDA in the smoke test, and we cleaned up a few slipped-ticks-due-to-rapid-tick-rate issues already), so it really shouldn't be misbehaving.

File: tests/kernel/timer/timer_behavior/src/main.c [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+	/* A measure of how wide the distribution is, ideal is 0 */
+	variance = variance / (double)CONFIG_TIMER_TEST_SAMPLES;
+
+	/* A measure of timer precision, ideal is 0 */
+	double stddev = sqrtf(variance);
+
+	/* Use double precision math here as integer overflows are possible in doing all the
+	 * conversions otherwise
+	 */
```
Comment (peter-mitsis) [inline tests/kernel/timer/timer_behavior/src/main.c:82]: Disclaimer: I've not delved into the math here to determine how much of an effect (if any) the following is going to have.

I am always wary of floating point in cases of conversion such as this. Yes, they allow for larger numbers to be represented, but at the cost of accuracy.  uint64_t's obviously give us 64 bits to work with. Doubles however only give us 52 bits in the significand (and the implicit 1).

File: tests/kernel/timer/timer_behavior/src/main.c (hunk 3 above)
Comment (teburd) [reply tests/kernel/timer/timer_behavior/src/main.c]: That is really helpful to know, thank you!

I've dropped all the rollover differences in the stats, but without knowing the maximum count of the timer (is this Kconfig/dts defined somewhere? if not that would be nice) I don't think knowing the true total time elapsed will work in all cases. I'll try it on some more hardware next week.

**Comment by teburd** [comment]
> Some notes. I love the idea here. I'm curious about the intel_adsp issues you're seeing. The timer hardware has been banged pretty hard there (e.g. there's a calibration loop between CCOUNT and HDA in the smoke test, and we cleaned up a few slipped-ticks-due-to-rapid-tick-rate issues already), so it really shouldn't be misbehaving.

I'm glad you like it, its really trying to codify some expectations of how a should a kernel timer behave when given a fixed period.

**Comment by nashif** [comment]
> I'm curious about the intel_adsp issues you're seeing. 

A new variant of  the issues we had this spring and this one as well https://github.com/zephyrproject-rtos/zephyr/issues/46378

**Comment by teburd** [comment]
Minor tweak to ensure the first timer call fires after the expected set period not with K_NO_WAIT as initially setup, this explains the relatively large negative total time I saw before.

frdm_k64f now results in this which looks very nice


```
START - test_periodic_behavior
periodic timer behavior test
periodic timer samples gathered, calculating statistics
timer clock rate 120000000, kernel tick rate 10000, expected period 1000 us, samples 10000, period 120000.000000 cycles
statistics samples 10000, rollovers 0, mean 999.996883 us, variance 0.097387 us, stddev 0.312069 us, min 968.858333 us, max 1000.025000 us
statistics samples 10000, rollovers 0, mean 119999.626000 cycles, variance 1402.376124 cycles, stddev 37.448311 cycles, min 116263 cycles, max 120003 cycles
timer start cycle 2406068, end cycle 1202402328, total time 9999968.833333, expected time 10000000.000000, difference -31.166667
 PASS - test_periodic_behavior in 10.211 seconds
```

**Review by ceolin — APPROVED** [review]

**Comment by teburd** [comment]
mimxrt1060_evk fails

```
START - test_periodic_behavior
periodic timer behavior test
periodic timer samples gathered, calculating statistics
timer clock rate 32768, kernel tick rate 4096, expected period 1000 us, samples 10000, period 32.768000 cycles
statistics samples 10000, rollovers 0, mean 1220.697021 us, variance 0.186227 us, stddev 0.431541 us, min 1190.185547 us, max 1220.703125 us
statistics samples 10000, rollovers 0, mean 39.999800 cycles, variance 0.000200 cycles, stddev 0.014141 cycles, min 39 cycles, max 40 cycles
timer start cycle 658, end cycle 400656, total time 12206970.214844, expected time 10000000.000000, difference 2206970.214844

    Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/timer/timer_behavior/src/main.c:166: timer_behavior_test_periodic_behavior: (max_us <= max_us_bound is false)
Longest timer period too long (off by more than 10%)
 FAIL - test_periodic_behavior in 12.291 seconds
```

**Comment by ceolin** [comment]
> mimxrt1060_evk fails
> 
> ```
> START - test_periodic_behavior
> periodic timer behavior test
> periodic timer samples gathered, calculating statistics
> timer clock rate 32768, kernel tick rate 4096, expected period 1000 us, samples 10000, period 32.768000 cycles
> statistics samples 10000, rollovers 0, mean 1220.697021 us, variance 0.186227 us, stddev 0.431541 us, min 1190.185547 us, max 1220.703125 us
> statistics samples 10000, rollovers 0, mean 39.999800 cycles, variance 0.000200 cycles, stddev 0.014141 cycles, min 39 cycles, max 40 cycles
> timer start cycle 658, end cycle 400656, total time 12206970.214844, expected time 10000000.000000, difference 2206970.214844
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/timer/timer_behavior/src/main.c:166: timer_behavior_test_periodic_behavior: (max_us <= max_us_bound is false)
> Longest timer period too long (off by more than 10%)
>  FAIL - test_periodic_behavior in 12.291 seconds
> ```

systematically ?

**Review by peter-mitsis — APPROVED** [review]

**Comment by carlescufi** [comment]
@andyross can you take a look?

**Review by nashif — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Refresh +1.  Sorry, fell a few days behind.

**Pull request merged**